src/config/builder.rs

```diff
@@ -63,6 +63,58 @@ struct ConfigBuilderHash<'a> {
     secret: BTreeMap<&'a ComponentKey, &'a dyn SecretBackend>,
 }
 
+#[cfg(feature = "enterprise")]
+impl ConfigBuilderHash<'_> {
+    /// Sort inner JSON values to maintain a consistent ordering. This prevents
+    /// non-deterministically serializable structures like HashMap from
+    /// affecting the resulting hash. As a consequence, ordering that does not
+    /// affect the actual semantics of a configuration is not considered when
+    /// calculating the hash.
+    fn into_hash(self) -> String {
+        use sha2::{Digest, Sha256};
+
+        let value = to_sorted_json_string(self);
+        let output = Sha256::digest(value.as_bytes());
+
+        hex::encode(output)
+    }
+}
+
+/// Converting to Value prior to serializing to JSON string is sufficient
+/// to sort our underlying keys. This is because Value::Map is backed (by
+/// default) by BTreeMap which maintains an implicit key order.
+/// Serializing Value (based on Map) is thus deterministic versus
+/// serializing ConfigBuilderHash (based on potential HashMap) directly.
+#[cfg(feature = "enterprise")]
+fn to_sorted_json_string<T>(value: T) -> String
+where
+    T: Serialize,
+{
+    let value = serde_json::to_value(value).expect("Should serialize to JSON. Please report.");
```

Comment:
Nice! I expected we'd need to traverse the keys manually. This is a welcome shortcut.

Reply:
Unfortunately, in looking into #13359 (see https://github.com/vectordotdev/vector/issues/13359#issuecomment-1168792560), it's become clear that some dependencies require `serde` with the `preserve_order` feature (`IndexMap` is used instead of `BTreeMap`) which then breaks our assumptions and logic. I decided to refactor this to an explicit sort which will be better long term. It was fun while it lasted 😅

Reply:
Nicely caught, though! `preserve_order` actually tripped me up in a historic issue. It was a good reminder not to rely on external implementation details, although I clearly forgot that fact here.

Reply:
Yep, lesson learned. Funnily enough, I actually had the explicit sort as my first attempt and then had an (in retrospect misguided) epiphany about `Value` and now here we are back again, though a little wiser for it.